### ISSUE TITLE
subworkflow order enforcing

### ISSUE THREAD

**Issue by DrYak** [snakemake/snakemake#743]
**Is your feature request related to a problem? Please describe.**

We have use cases where several sub-workflows need to be chained together, in our case: a first sub-workflow for downloading and preformatting sequencing data, another to process the data and find variants, and the finally the main workflows analyses the result of the later. The downloading is specific to this experiment (other use of the processing pipeline could uses other sources) and thus it cannot be made into a subworkflow of _that_ subworkflow.

Current [documentation of Snakemake state explicitely that all sub-workflow are executed prior to the main workflow](https://snakemake.readthedocs.io/en/stable/snakefiles/modularization.html#sub-workflows):
>  A sub-workflow is executed independently before the current workflow is executed. 

Thus it is not possible to enforce some order, e.g.: using rules in the main workflow.

It turns out that snakemake stores its subworkflow in a dictionnary, and [starting with Python 3.7 according to documentation](https://docs.python.org/3.7/library/stdtypes.html#typesmapping):
> Changed in version 3.7: Dictionary order is guaranteed to be insertion order. This behavior was an implementation detail of CPython from 3.6.

Thus it happens that the order in which subworkflow are mentionned in the main workflow will influence their execution order, so we could have a temporary workaround solution working for us, BUT...

**Describe the solution you'd like**

...but, the above ordering happens to be a lucky accident of the peculiar implementation of snakemake (dictionaries) and the peculiar platform it's currently running on (python >= 3.6).

It would be better with snakemake could have some `subworkfloworder:` directive, that either:
 - gave an explicit order, and snakemake runs first these subworkflow in that order, and then any remaining subworkflow
 - gave a list of `>` compared elements, that could then in turn be used to guide a `sort` of the keys (I would have said _a custom comparing function in a sort_ using [cmp_to_key python 3 adapter](https://docs.python.org/3/library/functools.html#functools.cmp_to_key) )

**Describe alternatives you've considered**

As mentionned, it happens to accidentally work in the current implementation, but it's relying on a happy coincidence that might get broken in future implementations of snakemake.
Having an official subworkflow reordering would make it easier.

**Additional context**

We're currently a bit busy with research into the current pandemic, so there won't be pull-request with suggestions for implementing that feature soon. Please be patient.

**Comment by nhartwic** [comment]
We already have a system for ordering execution. The entire point of snakemake is that it builds a dag based on dependencies and targets. Is there any reason sub workflows can't be incorporated into the DAG directly rather than trying to create a new system to manage how subworkflows are executed?

**Comment by DrYak** [comment]
In our case, the "main processing" mentioned in the middle, is an entire external pipeline, whereas the "downloading" and final "result analysis" are specific to a small project (for an article publication).
It's therefore better from our point of view to use sub-workflows rather than extend the main processing pipeline.

The "main processing pipeline" also builds it's DAG based on a list of samples (using snakemake's wildcards). Before the "download" completes, the samples aren't even there yet.
So making the full giant DAG of all sub-workflow combined isn't possible (at least it would require re-building the DAG after the download sub-workflow completes).

**Comment by nhartwic** [comment]
I'm just trying to point out that from a design perspective, snakemake already has checkpoints, dag building, and priority to solve ordering execution of tasks. Building another system for managing workflows seems inelegant.

The thing snakemake really lacks is a way to integrate a sub-workflow into these existing systems. I'm not saying you should have to extend the main pipeline. I'm saying that integrating subworkflows into the DAG (with all the associated control tools) is a better system for managing sub workflow execution than building a new priority system.

**Comment by kpj** [comment]
@nhartwic: I agree that duplicating all these features (checkpoints, DAG building, priority order) for subworkflows doesn't seem sensible. However, the current way subworkflows are handled seems somewhat limited (as you also mentioned).

See e.g. #679, #513 (maybe also #206, #119) for related issues. In addition, the DAG visualization does not gracefully include subworkflows, and "A sub-workflow is executed independently before the current workflow is executed" ([source](https://snakemake.readthedocs.io/en/stable/snakefiles/modularization.html#sub-workflows)) makes using them feel inflexible.

Are you aware of any efforts for (or at least the desire to) extending their functionality?

**Comment by nhartwic** [comment]
Well, as far as desires go, each of those issues you linked (I made one of them) is expressing a desire for cleaner integration of subworkflows. I haven't yet made a substantial effort to implement such a system. I imagine the user would write something like...

```
rule myotherworkflow:
    input:
        # some paths
    output:
        # some paths
    params:
        # specify workflow args here using dict like key-value pairs. alternative a function (f(wildcards) -> dict) that can be unpacked into key value pairs
    subworkflow:
        # path to other workflows snakefile
```

This rule could then be incorporated in the dag as normal, targets can be used as normal. And snakemake just needs to be told how to execute subworkflow-rules, analagous to how script-rules and shell-rules already exist.

**Comment by kpj** [comment]
That sounds like an interesting implementation suggestion!

I linked this issue on the Discord channel for further reach as well.
If there's enough support, one might move forward in this direction.

**Comment by johanneskoester** [comment]
> Well, as far as desires go, each of those issues you linked (I made one of them) is expressing a desire for cleaner integration of subworkflows. I haven't yet made a substantial effort to implement such a system. I imagine the user would write something like...
> 
> ```
> rule myotherworkflow:
>     input:
>         # some paths
>     output:
>         # some paths
>     params:
>         # specify workflow args here using dict like key-value pairs. alternative a function (f(wildcards) -> dict) that can be unpacked into key value pairs
>     subworkflow:
>         # path to other workflows snakefile
> ```
> 
> This rule could then be incorporated in the dag as normal, targets can be used as normal. And snakemake just needs to be told how to execute subworkflow-rules, analagous to how script-rules and shell-rules already exist. Just like with subworkflows now, arguments for the parent workflow should be passed to child workflows, with the exception of arguments specified in params which would be overwritten.

Hi all. This sounds indeed like an idea worth considering. I would be open to a tighter integration of subworkflows, either via the current syntax or via the one you suggest here. In any case, it should be done in a way that does not somehow duplicate the codebase. Integration of the subworkflow DAG into the main dag should be technically unproblematic though. The current implementation was just the simplest solution to get them working.

**Comment by kpj** [comment]
That sounds great!

With the above notation, one could try to simply merge the main-workflow DAG with the sub-workflow DAG (i.e. expand the sub-workflow node) during DAG generation time. This could potentially solve all (most?) current limitations. What do you think?

I'd be happy to give this a shot. @nhartwic: Would you also be interested? :)

**Comment by nhartwic** [comment]
I'm interested. I probably won't have 20 hours to spend on this till Christmas though. 

Whatever we decide as to DAG expansion, we need...
1. the config for the subworkflow to be independent from the config for the parent workflow
2. "--directory" option for the subworkflow should be independent from the parent workflow and should support wildcards
3. the config for the subworkflow should able to include wildcards. (or else be generated from a function f(wcs) -> dict )

...Everything else should be fine. But the '--config' and '--directory' options directly control where the workflow generates output so has to sink up with the 'output' fields of the subworkflow rules which may use wildcards.

Personally, I prefer my proposed syntax for subworkflow-rules since it matches the style of other rules. It is possible that they are different enough that they SHOULD look different syntactically though I'm hopeful that by making them more rule like and incorporating them into the dag, they will be easier for people to grok.

**Comment by kpj** [comment]
@nhartwic: I agree with all points you make. In a sense, they translate to continuing to support (and extend) the currently existing `workdir` and `configfile` directives for subworkflows.

I do also like the syntax you proposed quite a bit. The Snakefile will be specified using the `subworkflow` directive.
We could then think about whether specifying `workdir` and `configfile` could simply be done by setting the appropriate `params`.

I am quite unfamiliar with these parts of Snakemake's codebase, so this will be quite the adventure for me :)
Nonetheless, I started working here: https://github.com/kpj/snakemake/tree/feature_subworkflows (let me know if I should add you). In addition, I also created a draft PR: https://github.com/snakemake/snakemake/pull/791.

**Comment by kpj** [comment]
I have now implemented a possible (and so far very naive) solution in #791, and added some comments.
Let me know what you think :)
